pkg/engine/engine.go

```diff
@@ -0,0 +1,347 @@
+package engine
+
+import (
+	"errors"
+	"fmt"
+	"io"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"slices"
+	"strings"
+	"syscall"
+)
+
+var (
+	SupportedEngines = []string{"podman", "docker"}
+)
+
+type Container struct {
+	ID  string
+	Ref ContainerRef
+}
+
+type ContainerImage struct {
+	Registry   string
+	Repository string
+	Name       string
+	Tag        string
+	Hash       string
+}
+
+type ContainerRef struct {
+	Image          ContainerImage
+	Tag            string
+	Volumes        []VolumeMount
+	Envs           map[string]string
+	Tty            bool
+	PublishAll     bool
+	Interactive    bool
+	Entrypoint     string
+	Command        string
+	BestEffortArgs []string
+}
+
+type VolumeMount struct {
+	Source       string
+	Destination  string
+	MountOptions string
+}
+
+type Engine struct {
+	engine  string
+	binary  string
+	dryRun  bool
+	verbose bool
+}
+
+func (c ContainerRef) ImageFQDN() string {
+	i := fmt.Sprintf("%s:%s", c.Image.Name, c.Image.Tag)
+
+	// The order of the repository and registry addition is important
+	if c.Image.Repository != "" {
+		i = fmt.Sprintf("%s/%s", c.Image.Repository, i)
+	}
+
+	if c.Image.Registry != "" {
+		i = fmt.Sprintf("%s/%s", c.Image.Registry, i)
+	}
+
+	return i
+}
+
+func New(engine string, dryRun, verbose bool) (*Engine, error) {
+	e := &Engine{
+		dryRun:  dryRun,
+		verbose: verbose,
+	}
+
+	if verbose {
+		fmt.Println("using container engine:", engine)
+	}
```

Comment:
I'd like to put stuff like this behind a logging library, if possible.

Reply:
Fixed by 75364708cf18098746a0ea7205de0a2e38f6f16f